File: frontend/src/types/media/recommendations.ts
import { CampaignLifecycleState, MetricContract, RegionRecommendationRef, RegionTooltipData, WorkflowStatus } from './shared';

export interface CampaignBudgetPreview {
  weekly_budget_eur?: number;
  shift_pct?: number;
  shift_value_eur?: number;
  total_flight_budget_eur?: number;
}

export interface CampaignPreview {
  campaign_name?: string;
  activation_window?: {
    start?: string | null;
    end?: string | null;
  };
  budget?: CampaignBudgetPreview;
  primary_kpi?: string;
  recommended_product?: string;
  mapping_status?: string;
  playbook_key?: string;
  playbook_title?: string;
  ai_generation_status?: string;
}

export interface DecisionFact {
  key: string;
  label: string;
  value: string | number | boolean | null;
  source?: string;
}

export interface DecisionExpectation {
  condition_key?: string;
  condition_label?: string;
  region_codes?: string[];
  impact_probability?: number;
  signal_score?: number;
  peix_score?: number;
  signal_confidence_pct?: number | null;
  confidence_pct?: number;
  event_probability_pct?: number | null;
  field_contracts?: Record<string, MetricContract>;
  rationale?: string;
}

export interface DecisionRecommendation {
  primary_product?: string;
  primary_region?: string;
  secondary_regions?: string[];
  secondary_products?: string[];
  budget_shift_pct?: number;
  mapping_status?: string;
  mapping_reason?: string;
  action_required?: 'review_mapping' | 'ready_for_activation' | string;
}

export interface RecommendationDecisionBrief {
  summary_sentence?: string;
  horizon?: {
    min_days?: number;
    max_days?: number;
    model_lead_time_days?: number | null;
  };
  facts?: DecisionFact[];
  expectation?: DecisionExpectation;
  recommendation?: DecisionRecommendation;
}

export interface RecommendationCard {
  id: string;
  status: WorkflowStatus | string;
  status_label?: string;
  evidence_class?: string;
  type: string;
  urgency_score: number;
  brand: string;
  product: string;
  region?: string;
  region_codes?: string[];
  region_codes_display?: string[];
  display_title?: string;
  budget_shift_pct: number;
  channel_mix: Record<string, number>;
  activation_window?: {
    start?: string | null;
    end?: string | null;
  };
  reason?: string;
  confidence?: number;
  signal_score?: number;
  priority_score?: number;
  signal_confidence_pct?: number | null;
  outcome_signal_score?: number | null;
  outcome_confidence_pct?: number | null;
  learning_state?: string;
  outcome_learning_scope?: string;
  outcome_learning_explanation?: string;
  observed_response?: {
    response_units?: number | null;
    response_per_1000_eur?: number | null;
    qualified_visits?: number | null;
    avg_search_lift_index?: number | null;
  } | null;
  learned_lifts?: Array<{ label: string; value?: number | null }>;
  field_contracts?: Record<string, MetricContract>;
  campaign_name?: string;
  primary_kpi?: string;
  recommended_product?: string;
  mapping_status?: string;
  mapping_confidence?: number | null;
  mapping_reason?: string;
  mapping_rule_source?: string | null;
  condition_key?: string;
  condition_label?: string;
  mapping_candidate_product?: string | null;
  peix_context?: {
    region_code?: string;
    score?: number;
    signal_score?: number;
    band?: string;
    impact_probability?: number;
    drivers?: Array<{ label: string; strength_pct: number }>;
    trigger_event?: string;
  };
  playbook_key?: string;
  playbook_title?: string;
  trigger_snapshot?: {
    source?: string;
    event?: string;
    details?: string;
    lead_time_days?: number;
    values?: Record<string, number | string | boolean>;
  };
  guardrail_notes?: string[];
  ai_generation_status?: string;
  strategy_mode?: string;
  campaign_preview?: CampaignPreview;
  decision_brief?: RecommendationDecisionBrief;
  detail_url?: string;
  created_at?: string | null;
  updated_at?: string | null;
  expires_at?: string | null;
  is_supply_gap_active?: boolean;
  supply_gap_match_examples?: string;
  recommended_priority_multiplier?: number;
  supply_gap_product?: string;
  lifecycle_state?: CampaignLifecycleState | string;
  freshness_state?: 'scheduled' | 'current' | 'missing_window' | 'expired' | 'stale' | string;
  evidence_strength?: 'hoch' | 'mittel' | 'niedrig' | string;
  publish_blockers?: string[];
  is_publishable?: boolean;
  dedupe_group_id?: string;
  is_primary_variant?: boolean;
  decision_link?: string;
  variant_count?: number;
  variants?: Array<{
    id: string;
    status?: string;
    lifecycle_state?: CampaignLifecycleState | string;
    display_title?: string;
  }>;
}

export interface CampaignChannelPlanItem {
  channel: string;
  role: string;
  share_pct: number;
  budget_eur?: number;
  formats?: string[];
  message_angle?: string;
  kpi_primary?: string;
  kpi_secondary?: string[];
}

export interface CampaignPack {
  meta?: {
    version?: string;
    generated_at?: string;
    generator?: string;
  };
  campaign?: {
    campaign_name?: string;
    objective?: string;
    status?: WorkflowStatus | string;
    priority?: string;
  };
  targeting?: {
    region_scope?: string;
    audience_segments?: string[];
  };
  activation_window?: {
    start?: string;
    end?: string;
    flight_days?: number;
  };
  budget_plan?: {
    weekly_budget_eur?: number;
    budget_shift_pct?: number;
    budget_shift_value_eur?: number;
    total_flight_budget_eur?: number;
    currency?: string;
  };
  channel_plan?: CampaignChannelPlanItem[];
  message_framework?: {
    hero_message?: string;
    support_points?: string[];
    compliance_note?: string;
    cta?: string;
    copy_status?: string;
    library_version?: string;
    library_source?: string;
  };
  playbook?: {
    key?: string;
    title?: string;
    kind?: string;
    message_direction?: string;
    condition_key?: string;
  };
  trigger_snapshot?: {
    source?: string;
    event?: string;
    details?: string;
    lead_time_days?: number;
    values?: Record<string, number | string | boolean>;
  };
  trigger_evidence?: {
    source?: string;
    event?: string;
    details?: string;
    lead_time_days?: number;
    confidence?: number;
  };
  product_mapping?: {
    recommended_product?: string;
    mapping_status?: string;
    mapping_confidence?: number | null;
    mapping_reason?: string;
    condition_key?: string;
    condition_label?: string;
    candidate_product?: string | null;
  };
  measurement_plan?: {
    primary_kpi?: string;
    secondary_kpis?: string[];
    reporting_cadence?: string;
    success_criteria?: string;
  };
  peix_context?: {
    region_code?: string;
    score?: number;
    band?: string;
    impact_probability?: number;
    drivers?: Array<{ label: string; strength_pct: number }>;
    trigger_event?: string;
  };
  ai_plan?: {
    campaign_name?: string;
    objective?: string;
    budget_shift_pct?: number;
    activation_window_days?: number;
    channel_plan?: CampaignChannelPlanItem[];
    keyword_clusters?: string[];
    creative_angles?: string[];
    kpi_targets?: {
      primary_kpi?: string;
      secondary_kpis?: string[];
      success_criteria?: string;
    };
    next_steps?: Array<{
      task?: string;
      owner?: string;
      eta?: string;
    }>;
    compliance_hinweis?: string;
  };
  guardrail_report?: {
    passed?: boolean;
    notes?: string[];
    applied_fixes?: string[];
  };
  ai_meta?: {
    generated_at?: string;
    model?: string;
    provider?: string;
    status?: string;
    fallback_used?: boolean;
    error?: string;
  };
  execution_checklist?: Array<{
    task?: string;
    owner?: string;
    eta?: string;
    status?: string;
  }>;
}

export interface RecommendationDetail extends RecommendationCard {
  campaign_pack: CampaignPack;
  trigger_evidence?: CampaignPack['trigger_evidence'];
  target_audience?: string[];
  decision_brief?: RecommendationDecisionBrief;
}

export interface ConnectorCatalogItem {
  key: string;
  label: string;
  status: string;
  description?: string;
  supported_channels?: string[];
  supported_objectives?: string[];
}

export interface PreparedSyncPayload {
  opportunity_id: string;
  connector_key: string;
  connector_label: string;
  generated_at: string;
  available_connectors: ConnectorCatalogItem[];
  readiness: {
    state: 'ready' | 'approval_required' | 'needs_work' | string;
    can_sync_now: boolean;
    blockers: string[];
    warnings: string[];
    connector_status?: string;
  };
  normalized_package: {
    campaign_name?: string;
    workflow_status?: string;
    brand?: string;
    objective?: string;
    recommended_product?: string;
    region_codes?: string[];
    region_labels?: string[];
    audience_segments?: string[];
    primary_kpi?: string;
    secondary_kpis?: string[];
    budget_plan?: {
      weekly_budget_eur?: number;
      budget_shift_pct?: number;
      budget_shift_value_eur?: number;
      total_flight_budget_eur?: number;
      currency?: string;
    };
    activation_window?: {
      start?: string;
      end?: string;
      flight_days?: number;
    };
    channel_plan?: CampaignChannelPlanItem[];
    message_framework?: {
      hero_message?: string;
      support_points?: string[];
      cta?: string;
      compliance_note?: string;
    };
    creative_angles?: string[];
    keyword_clusters?: string[];
    next_steps?: Array<Record<string, unknown>>;
    playbook?: {
      key?: string;
      title?: string;
      kind?: string;
    };
    trigger?: {
      source?: string;
      event?: string;
      details?: string;
      lead_time_days?: number;
    };
    guardrails?: {
      passed?: boolean;
      notes?: string[];
      applied_fixes?: string[];
    };
  };
  connector_payload: Record<string, unknown>;
}

File: backend/app/services/media/cockpit/recommendations.py
from __future__ import annotations

from typing import Any

from sqlalchemy.orm import Session

from app.models.database import MarketingOpportunity
from app.services.media.cockpit.constants import BUNDESLAND_NAMES, REGION_NAME_TO_CODE
from app.services.media.recommendation_contracts import to_card_response
from app.services.media.semantic_contracts import (
    normalize_confidence_pct,
    signal_confidence_contract,
)

LEGACY_TO_WORKFLOW = {
    "NEW": "DRAFT",
    "URGENT": "DRAFT",
    "SENT": "APPROVED",
    "CONVERTED": "ACTIVATED",
}


def build_recommendation_section(db: Session) -> dict[str, Any]:
    rows = db.query(MarketingOpportunity).order_by(
        MarketingOpportunity.urgency_score.desc(),
        MarketingOpportunity.created_at.desc(),
    ).limit(20).all()

    cards: list[dict[str, Any]] = []
    for row in rows:
        region_codes = extract_region_codes_from_row(row)
        primary_region = region_codes[0] if region_codes else "Gesamt"
        channel_mix = row.channel_mix or {"programmatic": 35, "social": 30, "search": 20, "ctv": 15}
        campaign_payload = row.campaign_payload or {}
        campaign = campaign_payload.get("campaign") or {}
        budget = campaign_payload.get("budget_plan") or {}
        measurement = campaign_payload.get("measurement_plan") or {}
        activation = campaign_payload.get("activation_window") or {}
        product_mapping = campaign_payload.get("product_mapping") or {}
        peix_context = campaign_payload.get("peix_context") or {}
        playbook = campaign_payload.get("playbook") or {}
        ai_meta = campaign_payload.get("ai_meta") or {}
        status = LEGACY_TO_WORKFLOW.get(str(row.status or "").upper(), row.status or "DRAFT")
        recommended_product = product_mapping.get("recommended_product") or row.product or "Atemwegslinie"
        decision_expectation = (campaign_payload.get("decision_brief") or {}).get("expectation") or {}
        signal_confidence_pct = (
            normalize_confidence_pct(decision_expectation.get("signal_confidence_pct"))
            or normalize_confidence_pct(decision_expectation.get("confidence_pct"))
            or normalize_confidence_pct(
                ((campaign_payload.get("forecast_assessment") or {}).get("event_forecast") or {}).get("confidence")
            )
        )
        signal_score = (
            _coerce_float((peix_context or {}).get("signal_score"))
            or _coerce_float((peix_context or {}).get("score"))
            or _coerce_float((peix_context or {}).get("impact_probability"))
        )
        opp_payload = {
            "id": row.opportunity_id,
            "status": status,
            "type": row.opportunity_type,
            "urgency_score": row.urgency_score,
            "priority_score": row.urgency_score,
            "brand": row.brand or "PEIX Partner",
            "product": recommended_product,
            "recommended_product": recommended_product,
            "region": primary_region,
            "region_codes": region_codes,
            "budget_shift_pct": row.budget_shift_pct if row.budget_shift_pct is not None else 15.0,
            "channel_mix": channel_mix,
            "activation_window": {
                "start": (
                    row.activation_start.isoformat() if row.activation_start
                    else activation.get("start")
                ),
                "end": (
                    row.activation_end.isoformat() if row.activation_end
                    else activation.get("end")
                ),
            },
            "recommendation_reason": row.recommendation_reason or (row.trigger_event or "Epidemiologisches Trigger-Signal"),
            "confidence": (
                round(float(signal_confidence_pct) / 100.0, 2)
                if signal_confidence_pct is not None
                else None
            ),
            "signal_score": signal_score,
            "signal_confidence_pct": signal_confidence_pct,
            "mapping_status": product_mapping.get("mapping_status"),
            "mapping_confidence": product_mapping.get("mapping_confidence"),
            "mapping_reason": product_mapping.get("mapping_reason"),
            "condition_key": product_mapping.get("condition_key"),
            "condition_label": product_mapping.get("condition_label"),
            "mapping_candidate_product": product_mapping.get("candidate_product"),
            "playbook_key": row.playbook_key or playbook.get("key"),
            "playbook_title": playbook.get("title"),
            "strategy_mode": row.strategy_mode or campaign_payload.get("strategy_mode"),
            "trigger_snapshot": campaign_payload.get("trigger_snapshot"),
            "guardrail_notes": (campaign_payload.get("guardrail_report") or {}).get("applied_fixes") or [],
            "ai_generation_status": ai_meta.get("status"),
            "campaign_name": campaign.get("campaign_name"),
            "primary_kpi": measurement.get("primary_kpi"),
            "peix_context": peix_context,
            "campaign_payload": campaign_payload,
            "campaign_preview": {
                "campaign_name": campaign.get("campaign_name"),
                "activation_window": {
                    "start": activation.get("start"),
                    "end": activation.get("end"),
                },
                "budget": {
                    "weekly_budget_eur": budget.get("weekly_budget_eur"),
                    "shift_pct": budget.get("budget_shift_pct"),
                    "shift_value_eur": budget.get("budget_shift_value_eur"),
                    "total_flight_budget_eur": budget.get("total_flight_budget_eur"),
                },
                "primary_kpi": measurement.get("primary_kpi"),
                "recommended_product": recommended_product,
                "mapping_status": product_mapping.get("mapping_status"),
            },
            "detail_url": f"/kampagnen/{row.opportunity_id}",
            "created_at": row.created_at.isoformat() if row.created_at else None,
            "updated_at": row.updated_at.isoformat() if row.updated_at else None,
        }
        card = to_card_response(opp_payload, include_preview=True)
        card.setdefault("field_contracts", {})
        card["field_contracts"]["signal_confidence_pct"] = signal_confidence_contract(
            source=str(
                (campaign_payload.get("trigger_snapshot") or {}).get("source")
                or "Signal-Fusion"
            ),
            derived_from="trigger_evidence.confidence",
        )
        cards.append(card)

    return {
        "total": len(cards),
        "cards": cards,
    }


def build_region_recommendation_refs(db: Session) -> dict[str, dict[str, Any]]:
    refs: dict[str, dict[str, Any]] = {}
    rows = db.query(MarketingOpportunity).order_by(
        MarketingOpportunity.urgency_score.desc(),
        MarketingOpportunity.created_at.desc(),
    ).limit(250).all()

    for row in rows:
        status = LEGACY_TO_WORKFLOW.get(str(row.status or "").upper(), str(row.status or "DRAFT").upper())
        if status in {"DISMISSED", "EXPIRED"}:
            continue

        region_codes = extract_region_codes_from_row(row)
        if not region_codes:
            region_codes = sorted(BUNDESLAND_NAMES.keys())

        payload = {
            "card_id": row.opportunity_id,
            "detail_url": f"/kampagnen/{row.opportunity_id}",
            "status": status,
            "urgency_score": row.urgency_score,
            "priority_score": row.urgency_score,
            "brand": row.brand,
            "product": row.product,
        }

        for code in region_codes:
            if code not in refs:
                refs[code] = payload

    return refs


def extract_region_codes_from_row(row: MarketingOpportunity) -> list[str]:
    region_target = row.region_target or {}
    campaign_payload = row.campaign_payload or {}
    targeting = campaign_payload.get("targeting") or {}

    tokens: list[str] = []
    states = region_target.get("states")
    if isinstance(states, list):
        tokens.extend(str(item) for item in states)

    scope = targeting.get("region_scope")
    if isinstance(scope, list):
        tokens.extend(str(item) for item in scope)
    elif isinstance(scope, str):
        tokens.append(scope)

    normalized: set[str] = set()
    for token in tokens:
        lower = token.strip().lower()
        if lower in {"gesamt", "all", "de", "national", "deutschland"}:
            return []

        code = token.strip().upper()
        if code in BUNDESLAND_NAMES:
            normalized.add(code)
            continue

        mapped = REGION_NAME_TO_CODE.get(lower)
        if mapped:
            normalized.add(mapped)

    return sorted(normalized)


def _coerce_float(value: Any) -> float | None:
    if value is None:
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None
